README.md

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/JohnCoene/countup.svg?branch=master)](https://travis-ci.org/JohnCoene/countup)
<!-- badges: end -->

![](./man/figures/logo.png)

The underlying JavaScript library is
[CountUp.js](https://github.com/inorganik/CountUp.js).

## Install

Install from Github with `remotes`:

``` r
# install.packages("remotes")
remotes::install_github("JohnCoene/countup")
```

## News

Since version `0.1.1` the package also includes the [odometer](https://github.hubspot.com/odometer/docs/welcome/) widget.

```r
odometer(12345)
```

## Examples

In the R console or R markdown.

``` r
library(countup)

countup(25)

opts = list(
  useEasing = TRUE, 
  useGrouping = TRUE, 
  separator = ',', 
  decimal = '.', 
  prefix = 'US$', 
  suffix = ' bejillion' 
)

countup(count = 1729, start = 10, options = opts)
```

In Shiny (since version `0.1.0`), there are methods to programatically
interact with the counter.

``` r
library(shiny)
library(countup)

ui <- fluidPage(
  h1(countupOutput("cnt")),
  actionButton("start", "start"),
  actionButton("pause", "pause / resume"),
  numericInput("value", "update to:", min = 1, max = 1000, value = 24)
)

server <- function(input, output, session) {

  output$cnt <- renderCountup({
    countup(1000, duration = 10, start = FALSE)
  })

  observeEvent(input$start, {
    countupProxy("cnt") %>% 
      countup_start()
  })

  observeEvent(input$value, {
    countupProxy("cnt") %>% 
      countup_update(input$value)
  })

  observeEvent(input$pause, {
    countupProxy("cnt") %>% 
      countup_pause_resume()
  })

}

shinyApp(ui, server)
```


NEWS.md
# countup 0.1.1

Added [odometer](https://github.hubspot.com/odometer/), see [#3](https://github.com/JohnCoene/countup/issues/3).

# countup 0.1.0

In `countup`:

- The start argument is renamed to `start_at`.
- Addition of `start` argument to set whether to start the counter on draw (or programmatically).

Proxy:

- `counter_start` to programmatically start the counter.
- `counter_update` to programmatically update the counter.
- `counter_pause_resume` to programmatically pause or resume the counter.

# countup 0.0.1

* Initial Version


R/countup.R
#' countup
#'
#' Count up widget
#'
#' @param count Numeric or integer to count up to (required).
#' @param start Whether to start the counter.
#' @param start_at Integer to start from defaults to \code{0}.
#' @param options List of options (see details and examples).
#' @param duration duration of the count defaults to \code{2.5}.
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param elementId Specify element id of \code{<span>} (optional).
#'
#' @details Valid options include:
#' \itemize{
#' \item{useEasing }{takes \code{TRUE} or \code{FALSE}}
#' \item{useGrouping }{takes \code{TRUE} or \code{FALSE}}
#' \item{separator }{takes a string, i.e.: ","}
#' \item{decimal }{takes a string, i.e.: "."}
#' \item{prefix }{takes a string, i.e.: "$"}
#' \item{suffix }{takes a string, i.e.: "thousands"}
#' }
#'
#' @examples
#' \dontrun{
#' countup(25)
#'
#' opts <- list(useEasing = TRUE,
#'              useGrouping = TRUE,
#'              separator = "'",
#'              prefix = "US$",
#'              suffix = " million")
#'
#' countup(1234, start = 200, options = opts)
#' }
#' 
#' @return An object of class `countup`
#'
#' @import htmlwidgets
#'
#' @export
countup <- function(count, start_at = 0, options = NULL, duration = 2.5, 
                    start = TRUE, width = NULL, height = NULL, 
                    elementId = NULL) {

  if(missing(count)) stop("must pass count")

  # forward options using x
  x = list(
    count = count,
    start = start,
    start_at = start_at,
    options = options,
    duration = duration
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'countup',
    x,
    width = width,
    height = height,
    package = 'countup',
    elementId = elementId
  )
}

countup_html <- function(id, style, class, ...){
  htmltools::tags$span(id = id, class = class)
}

#' Shiny bindings for countup
#'
#' Output and render functions for using countup within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a countup
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#' @param id Id of \code{countup} object to create a proxy.
#' @param session A valid shiny session.
#'
#' @name countup-shiny
#'
#' @export
countupOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'countup', width, height, package = 'countup')
}

#' @rdname countup-shiny
#' @export
renderCountup <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, countupOutput, env, quoted = TRUE)
}

#' @rdname countup-shiny
#' @export
countupProxy <- function(id, session = shiny::getDefaultReactiveDomain()){
  
  proxy <- list(id = id, session = session)
  class(proxy) <- "countupProxy"
  
  return(proxy)
}

R/odometer.R
#' Odometer
#'
#' Create an odometer.
#' 
#' @param count Value to count up to.
#' @param duration Duration of countup.
#' @param format Change how digit groups are formatted, and 
#' how many digits are shown after the decimal point.
#' @param theme An odometer theme, see \url{https://github.hubspot.com/odometer/api/themes/}
#' @param width,height Dimensions.
#' @param elementId Id of element holding the odometer.
#' 
#' @examples 
#' odometer(12345)
#'
#' @import htmlwidgets
#'
#' @export
odometer <- function(count, duration = 2500, format = "(,ddd).dd",
  theme = c("default", "car", "digital", "minimal", "plaza", "slot-machine", "train-station"),
  width = NULL, height = NULL, elementId = NULL) {

  theme <- match.arg(theme)

  # forward options using x
  x = list(
    value = count,
    opts = list(
      duration = duration,
      theme = theme,
      format = format
    )
  )

  # create dependency
  path <- system.file("htmlwidgets/lib/odometer/themes", package = "countup")
  dep <- htmltools::htmlDependency(
    name = "odometer-themes",
    version = "0.4.8",
    src = c(file = path),
    stylesheet = paste0("odometer-theme-", theme, ".css")
  )

  # create widget
  htmlwidgets::createWidget(
    name = 'odometer',
    x,
    width = width,
    height = height,
    package = 'countup',
    elementId = elementId,
    dependencies = list(dep),
    sizingPolicy = htmlwidgets::sizingPolicy(
      defaultWidth = "auto",
      defaultHeight = "auto",
      padding = 0
    )
  )
}

odometer_html <- function(..., height, width, style){
  htmltools::tags$span(...)
}

#' Shiny bindings for odometer
#'
#' Output and render functions for using odometer within Shiny
#' applications and interactive Rmd documents.
#'
#' @param outputId output variable to read from
#' @param width,height Must be a valid CSS unit (like \code{'100\%'},
#'   \code{'400px'}, \code{'auto'}) or a number, which will be coerced to a
#'   string and have \code{'px'} appended.
#' @param expr An expression that generates a odometer
#' @param env The environment in which to evaluate \code{expr}.
#' @param quoted Is \code{expr} a quoted expression (with \code{quote()})? This
#'   is useful if you want to save an expression in a variable.
#'
#' @name odometer-shiny
#'
#' @export
odometerOutput <- function(outputId, width = '100%', height = '400px'){
  htmlwidgets::shinyWidgetOutput(outputId, 'odometer', width, height, package = 'countup')
}

#' @rdname odometer-shiny
#' @export
renderOdometer <- function(expr, env = parent.frame(), quoted = FALSE) {
  if (!quoted) { expr <- substitute(expr) } # force quoted
  htmlwidgets::shinyRenderWidget(expr, odometerOutput, env, quoted = TRUE)
}


inst/htmlwidgets/countup.js
HTMLWidgets.widget({

  name: 'countup',

  type: 'output',

  factory: function(el, width, height) {

    var count;

    return {

      renderValue: function(x) {
        
        var options = x.options;
      
        count = new CountUp(el.id, x.start_at, x.count, 0, x.duration, options);
      
        if(x.start == true)
          count.start(); 

      },

      resize: function(width, height) {

      },

      getCounter: function(){
        return count;
      },

    };
  }
});

function get_counter(id, variable){

  var htmlWidgetsObj = HTMLWidgets.find("#" + id);

  var cnt;

  if (typeof htmlWidgetsObj != 'undefined') {
    cnt = htmlWidgetsObj.getCounter(variable);
  }

  return(cnt);
}

if (HTMLWidgets.shinyMode) {
  Shiny.addCustomMessageHandler('counter_start',
    function(data) {
      var counter = get_counter(data.id);
      if (typeof counter != 'undefined')
        counter.start()
      
  });

  Shiny.addCustomMessageHandler('counter_update',
    function(data) {
      var counter = get_counter(data.id);
      if (typeof counter != 'undefined')
        counter.update(data.update)
      
  });

  Shiny.addCustomMessageHandler('counter_pause_resume',
    function(data) {
      var counter = get_counter(data.id);
      if (typeof counter != 'undefined')
        counter.pauseResume()
      
  });
}



inst/htmlwidgets/lib/countUp-1.8.1/countUp-1.8.1.js
/*

  countUp.js
by @inorganik

*/

  // target = id of html element or var of previously selected html element where counting occurs
// startVal = the value you want to begin at
// endVal = the value you want to arrive at
// decimals = number of decimal places, default 0
// duration = duration of animation in seconds, default 2
// options = optional object of options (see below)

var CountUp = function(target, startVal, endVal, decimals, duration, options) {

  // make sure requestAnimationFrame and cancelAnimationFrame are defined
  // polyfill for browsers without native support
  // by Opera engineer Erik Möller
  var lastTime = 0;
  var vendors = ['webkit', 'moz', 'ms', 'o'];
  for(var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
    window.requestAnimationFrame = window[vendors[x]+'RequestAnimationFrame'];
    window.cancelAnimationFrame =
      window[vendors[x]+'CancelAnimationFrame'] || window[vendors[x]+'CancelRequestAnimationFrame'];
  }
  if (!window.requestAnimationFrame) {
    window.requestAnimationFrame = function(callback, element) {
      var currTime = new Date().getTime();
      var timeToCall = Math.max(0, 16 - (currTime - lastTime));
      var id = window.setTimeout(function() { callback(currTime + timeToCall); },
                                 timeToCall);
      lastTime = currTime + timeToCall;
      return id;
    };
  }
  if (!window.cancelAnimationFrame) {
    window.cancelAnimationFrame = function(id) {
      clearTimeout(id);
    };
  }

  var self = this;

  // default options
  self.options = {
    useEasing : true, // toggle easing
    useGrouping : true, // 1,000,000 vs 1000000
    separator : ',', // character to use as a separator
    decimal : '.', // character to use as a decimal
    easingFn: null, // optional custom easing closure function, default is Robert Penner's easeOutExpo
    formattingFn: null // optional custom formatting function, default is self.formatNumber below
  };
    // extend default options with passed options object
    for (var key in options) {
    if (options.hasOwnProperty(key)) {
    self.options[key] = options[key];
    }
    }
    if (self.options.separator === '') { self.options.useGrouping = false; }
    if (!self.options.prefix) self.options.prefix = '';
    if (!self.options.suffix) self.options.suffix = '';

    self.d = (typeof target === 'string') ? document.getElementById(target) : target;
    self.startVal = Number(startVal);
    self.endVal = Number(endVal);
    self.countDown = (self.startVal > self.endVal);
    self.frameVal = self.startVal;
    self.decimals = Math.max(0, decimals || 0);
    self.dec = Math.pow(10, self.decimals);
    self.duration = Number(duration) * 1000 || 2000;

    self.formatNumber = function(nStr) {
    nStr = nStr.toFixed(self.decimals);
    nStr += '';
    var x, x1, x2, rgx;
    x = nStr.split('.');
    x1 = x[0];
    x2 = x.length > 1 ? self.options.decimal + x[1] : '';
    rgx = /(\d+)(\d{3})/;
    if (self.options.useGrouping) {
    while (rgx.test(x1)) {
    x1 = x1.replace(rgx, '$1' + self.options.separator + '$2');
    }
    }
    return self.options.prefix + x1 + x2 + self.options.suffix;
    };
    // Robert Penner's easeOutExpo
    self.easeOutExpo = function(t, b, c, d) {
      return c * (-Math.pow(2, -10 * t / d) + 1) * 1024 / 1023 + b;
    };

    self.easingFn = self.options.easingFn ? self.options.easingFn : self.easeOutExpo;
    self.formattingFn = self.options.formattingFn ? self.options.formattingFn : self.formatNumber;

    self.version = function () { return '1.8.1'; };

    // Print value to target
    self.printValue = function(value) {
      var result = self.formattingFn(value);

      if (self.d.tagName === 'INPUT') {
        this.d.value = result;
      }
      else if (self.d.tagName === 'text' || self.d.tagName === 'tspan') {
        this.d.textContent = result;
      }
      else {
        this.d.innerHTML = result;
      }
    };

    self.count = function(timestamp) {

      if (!self.startTime) { self.startTime = timestamp; }

      self.timestamp = timestamp;
      var progress = timestamp - self.startTime;
      self.remaining = self.duration - progress;

      // to ease or not to ease
      if (self.options.useEasing) {
        if (self.countDown) {
          self.frameVal = self.startVal - self.easingFn(progress, 0, self.startVal - self.endVal, self.duration);
        } else {
          self.frameVal = self.easingFn(progress, self.startVal, self.endVal - self.startVal, self.duration);
        }
      } else {
        if (self.countDown) {
          self.frameVal = self.startVal - ((self.startVal - self.endVal) * (progress / self.duration));
        } else {
          self.frameVal = self.startVal + (self.endVal - self.startVal) * (progress / self.duration);
        }
      }

      // don't go past endVal since progress can exceed duration in the last frame
      if (self.countDown) {
      self.frameVal = (self.frameVal < self.endVal) ? self.endVal : self.frameVal;
      } else {
      self.frameVal = (self.frameVal > self.endVal) ? self.endVal : self.frameVal;
      }

      // decimal
      self.frameVal = Math.round(self.frameVal*self.dec)/self.dec;

      // format and print value
      self.printValue(self.frameVal);

      // whether to continue
      if (progress < self.duration) {
      self.rAF = requestAnimationFrame(self.count);
      } else {
      if (self.callback) { self.callback(); }
      }
    };
      // start your animation
      self.start = function(callback) {
      self.callback = callback;
      self.rAF = requestAnimationFrame(self.count);
      return false;
      };
      // toggles pause/resume animation
      self.pauseResume = function() {
      if (!self.paused) {
      self.paused = true;
      cancelAnimationFrame(self.rAF);
      } else {
      self.paused = false;
      delete self.startTime;
      self.duration = self.remaining;
      self.startVal = self.frameVal;
      requestAnimationFrame(self.count);
      }
      };
      // reset to startVal so animation can be run again
      self.reset = function() {
      self.paused = false;
      delete self.startTime;
      self.startVal = startVal;
      cancelAnimationFrame(self.rAF);
      self.printValue(self.startVal);
      };
      // pass a new endVal and start animation
      self.update = function (newEndVal) {
      cancelAnimationFrame(self.rAF);
      self.paused = false;
      delete self.startTime;
      self.startVal = self.frameVal;
      self.endVal = Number(newEndVal);
      self.countDown = (self.startVal > self.endVal);
      self.rAF = requestAnimationFrame(self.count);
      };

      // format startVal on initialization
      self.printValue(self.startVal);
      };


inst/htmlwidgets/lib/waypoints-2.0.3/waypoints.min-2.0.3.js
// Generated by CoffeeScript 1.6.2
/*
  jQuery Waypoints - v2.0.3
Copyright (c) 2011-2013 Caleb Troughton
Dual licensed under the MIT license and GPL license.
https://github.com/imakewebthings/jquery-waypoints/blob/master/licenses.txt
*/
  (function(){var t=[].indexOf||function(t){for(var e=0,n=this.length;e<n;e++){if(e in this&&this[e]===t)return e}return-1},e=[].slice;(function(t,e){if(typeof define==="function"&&define.amd){return define("waypoints",["jquery"],function(n){return e(n,t)})}else{return e(t.jQuery,t)}})(this,function(n,r){var i,o,l,s,f,u,a,c,h,d,p,y,v,w,g,m;i=n(r);c=t.call(r,"ontouchstart")>=0;s={horizontal:{},vertical:{}};f=1;a={};u="waypoints-context-id";p="resize.waypoints";y="scroll.waypoints";v=1;w="waypoints-waypoint-ids";g="waypoint";m="waypoints";o=function(){function t(t){var e=this;this.$element=t;this.element=t[0];this.didResize=false;this.didScroll=false;this.id="context"+f++;this.oldScroll={x:t.scrollLeft(),y:t.scrollTop()};this.waypoints={horizontal:{},vertical:{}};t.data(u,this.id);a[this.id]=this;t.bind(y,function(){var t;if(!(e.didScroll||c)){e.didScroll=true;t=function(){e.doScroll();return e.didScroll=false};return r.setTimeout(t,n[m].settings.scrollThrottle)}});t.bind(p,function(){var t;if(!e.didResize){e.didResize=true;t=function(){n[m]("refresh");return e.didResize=false};return r.setTimeout(t,n[m].settings.resizeThrottle)}})}t.prototype.doScroll=function(){var t,e=this;t={horizontal:{newScroll:this.$element.scrollLeft(),oldScroll:this.oldScroll.x,forward:"right",backward:"left"},vertical:{newScroll:this.$element.scrollTop(),oldScroll:this.oldScroll.y,forward:"down",backward:"up"}};if(c&&(!t.vertical.oldScroll||!t.vertical.newScroll)){n[m]("refresh")}n.each(t,function(t,r){var i,o,l;l=[];o=r.newScroll>r.oldScroll;i=o?r.forward:r.backward;n.each(e.waypoints[t],function(t,e){var n,i;if(r.oldScroll<(n=e.offset)&&n<=r.newScroll){return l.push(e)}else if(r.newScroll<(i=e.offset)&&i<=r.oldScroll){return l.push(e)}});l.sort(function(t,e){return t.offset-e.offset});if(!o){l.reverse()}return n.each(l,function(t,e){if(e.options.continuous||t===l.length-1){return e.trigger([i])}})});return this.oldScroll={x:t.horizontal.newScroll,y:t.vertical.newScroll}};t.prototype.refresh=function(){var t,e,r,i=this;r=n.isWindow(this.element);e=this.$element.offset();this.doScroll();t={horizontal:{contextOffset:r?0:e.left,contextScroll:r?0:this.oldScroll.x,contextDimension:this.$element.width(),oldScroll:this.oldScroll.x,forward:"right",backward:"left",offsetProp:"left"},vertical:{contextOffset:r?0:e.top,contextScroll:r?0:this.oldScroll.y,contextDimension:r?n[m]("viewportHeight"):this.$element.height(),oldScroll:this.oldScroll.y,forward:"down",backward:"up",offsetProp:"top"}};return n.each(t,function(t,e){return n.each(i.waypoints[t],function(t,r){var i,o,l,s,f;i=r.options.offset;l=r.offset;o=n.isWindow(r.element)?0:r.$element.offset()[e.offsetProp];if(n.isFunction(i)){i=i.apply(r.element)}else if(typeof i==="string"){i=parseFloat(i);if(r.options.offset.indexOf("%")>-1){i=Math.ceil(e.contextDimension*i/100)}}r.offset=o-e.contextOffset+e.contextScroll-i;if(r.options.onlyOnScroll&&l!=null||!r.enabled){return}if(l!==null&&l<(s=e.oldScroll)&&s<=r.offset){return r.trigger([e.backward])}else if(l!==null&&l>(f=e.oldScroll)&&f>=r.offset){return r.trigger([e.forward])}else if(l===null&&e.oldScroll>=r.offset){return r.trigger([e.forward])}})})};t.prototype.checkEmpty=function(){if(n.isEmptyObject(this.waypoints.horizontal)&&n.isEmptyObject(this.waypoints.vertical)){this.$element.unbind([p,y].join(" "));return delete a[this.id]}};return t}();l=function(){function t(t,e,r){var i,o;r=n.extend({},n.fn[g].defaults,r);if(r.offset==="bottom-in-view"){r.offset=function(){var t;t=n[m]("viewportHeight");if(!n.isWindow(e.element)){t=e.$element.height()}return t-n(this).outerHeight()}}this.$element=t;this.element=t[0];this.axis=r.horizontal?"horizontal":"vertical";this.callback=r.handler;this.context=e;this.enabled=r.enabled;this.id="waypoints"+v++;this.offset=null;this.options=r;e.waypoints[this.axis][this.id]=this;s[this.axis][this.id]=this;i=(o=t.data(w))!=null?o:[];i.push(this.id);t.data(w,i)}t.prototype.trigger=function(t){if(!this.enabled){return}if(this.callback!=null){this.callback.apply(this.element,t)}if(this.options.triggerOnce){return this.destroy()}};t.prototype.disable=function(){return this.enabled=false};t.prototype.enable=function(){this.context.refresh();return this.enabled=true};t.prototype.destroy=function(){delete s[this.axis][this.id];delete this.context.waypoints[this.axis][this.id];return this.context.checkEmpty()};t.getWaypointsByElement=function(t){var e,r;r=n(t).data(w);if(!r){return[]}e=n.extend({},s.horizontal,s.vertical);return n.map(r,function(t){return e[t]})};return t}();d={init:function(t,e){var r;if(e==null){e={}}if((r=e.handler)==null){e.handler=t}this.each(function(){var t,r,i,s;t=n(this);i=(s=e.context)!=null?s:n.fn[g].defaults.context;if(!n.isWindow(i)){i=t.closest(i)}i=n(i);r=a[i.data(u)];if(!r){r=new o(i)}return new l(t,r,e)});n[m]("refresh");return this},disable:function(){return d._invoke(this,"disable")},enable:function(){return d._invoke(this,"enable")},destroy:function(){return d._invoke(this,"destroy")},prev:function(t,e){return d._traverse.call(this,t,e,function(t,e,n){if(e>0){return t.push(n[e-1])}})},next:function(t,e){return d._traverse.call(this,t,e,function(t,e,n){if(e<n.length-1){return t.push(n[e+1])}})},_traverse:function(t,e,i){var o,l;if(t==null){t="vertical"}if(e==null){e=r}l=h.aggregate(e);o=[];this.each(function(){var e;e=n.inArray(this,l[t]);return i(o,e,l[t])});return this.pushStack(o)},_invoke:function(t,e){t.each(function(){var t;t=l.getWaypointsByElement(this);return n.each(t,function(t,n){n[e]();return true})});return this}};n.fn[g]=function(){var t,r;r=arguments[0],t=2<=arguments.length?e.call(arguments,1):[];if(d[r]){return d[r].apply(this,t)}else if(n.isFunction(r)){return d.init.apply(this,arguments)}else if(n.isPlainObject(r)){return d.init.apply(this,[null,r])}else if(!r){return n.error("jQuery Waypoints needs a callback function or handler option.")}else{return n.error("The "+r+" method does not exist in jQuery Waypoints.")}};n.fn[g].defaults={context:r,continuous:true,enabled:true,horizontal:false,offset:0,triggerOnce:false};h={refresh:function(){return n.each(a,function(t,e){return e.refresh()})},viewportHeight:function(){var t;return(t=r.innerHeight)!=null?t:i.height()},aggregate:function(t){var e,r,i;e=s;if(t){e=(i=a[n(t).data(u)])!=null?i.waypoints:void 0}if(!e){return[]}r={horizontal:[],vertical:[]};n.each(r,function(t,i){n.each(e[t],function(t,e){return i.push(e)});i.sort(function(t,e){return t.offset-e.offset});r[t]=n.map(i,function(t){return t.element});return r[t]=n.unique(r[t])});return r},above:function(t){if(t==null){t=r}return h._filter(t,"vertical",function(t,e){return e.offset<=t.oldScroll.y})},below:function(t){if(t==null){t=r}return h._filter(t,"vertical",function(t,e){return e.offset>t.oldScroll.y})},left:function(t){if(t==null){t=r}return h._filter(t,"horizontal",function(t,e){return e.offset<=t.oldScroll.x})},right:function(t){if(t==null){t=r}return h._filter(t,"horizontal",function(t,e){return e.offset>t.oldScroll.x})},enable:function(){return h._invoke("enable")},disable:function(){return h._invoke("disable")},destroy:function(){return h._invoke("destroy")},extendFn:function(t,e){return d[t]=e},_invoke:function(t){var e;e=n.extend({},s.vertical,s.horizontal);return n.each(e,function(e,n){n[t]();return true})},_filter:function(t,e,r){var i,o;i=a[n(t).data(u)];if(!i){return[]}o=[];n.each(i.waypoints[e],function(t,e){if(r(i,e)){return o.push(e)}});o.sort(function(t,e){return t.offset-e.offset});return n.map(o,function(t){return t.element})}};n[m]=function(){var t,n;n=arguments[0],t=2<=arguments.length?e.call(arguments,1):[];if(h[n]){return h[n].apply(null,t)}else{return h.aggregate.call(null,n)}};n[m].settings={resizeThrottle:100,scrollThrottle:30};return i.load(function(){return n[m]("refresh")})})}).call(this);


R/proxy.R
#' Start
#' 
#' Start the counter.
#' 
#' @param counter An object of class `countupProxy` as returned by [countupProxy()].
#' 
#' @examples 
#' library(shiny)
#' 
#' ui <- fluidPage(
#'   countupOutput("cnt"),
#'   actionButton("start", "start")
#' )
#' 
#' server <- function(input, output, session) {
#' 
#'   output$cnt <- renderCountup({
#'     countup(24, start = FALSE)
#'   })
#' 
#'   observeEvent(input$start, {
#'     countupProxy("cnt") %>% 
#'       countup_start()
#'   })
#' 
#' }
#' 
#' \dontrun{shinyApp(ui, server)}
#' 
#' @return The `counter`
#' 
#' @export
countup_start <- function(counter) UseMethod("countup_start")

#' @export 
#' @method countup_start countupProxy
countup_start.countupProxy <- function(counter){
  counter$session$sendCustomMessage("counter_start", list(id = counter$id))
  invisible(counter)
}

#' Update
#' 
#' Update the counter.
#' 
#' @inheritParams countup_start
#' @param update Integer to update to.
#' 
#' @examples 
#' library(shiny)
#' 
#' ui <- fluidPage(
#'   countupOutput("cnt"),
#'   actionButton("Update", "start")
#' )
#' 
#' server <- function(input, output, session) {
#' 
#'   output$cnt <- renderCountup({
#'     countup(24)
#'   })
#' 
#'   observeEvent(input$start, {
#'     countupProxy("cnt") %>% 
#'       countup_update(50)
#'   })
#' 
#' }
#' 
#' \dontrun{shinyApp(ui, server)}
#' 
#' @return The `counter`
#' 
#' @export
countup_update <- function(counter, update) UseMethod("countup_update")

#' @export 
#' @method countup_update countupProxy
countup_update.countupProxy <- function(counter, update){
  counter$session$sendCustomMessage("counter_update", list(id = counter$id, update = update))
  invisible(counter)
}

#' Pause Resume
#' 
#' Pause or resume the counter.
#' 
#' @inheritParams countup_start
#' 
#' @examples 
#' library(shiny)
#' 
#' ui <- fluidPage(
#'   countupOutput("cnt"),
#'   actionButton("pause/resume", "pause")
#' )
#' 
#' server <- function(input, output, session) {
#' 
#'   output$cnt <- renderCountup({
#'     countup(1000, duration = 10)
#'   })
#' 
#'   observeEvent(input$pause, {
#'     countupProxy("cnt") %>% 
#'       countup_pause_resume()
#'   })
#' 
#' }
#' 
#' \dontrun{shinyApp(ui, server)}
#' 
#' @return The `counter`
#' 
#' @export
countup_pause_resume <- function(counter) UseMethod("countup_pause_resume")

#' @export 
#' @method countup_pause_resume countupProxy
countup_pause_resume.countupProxy <- function(counter){
  counter$session$sendCustomMessage("counter_pause_resume", list(id = counter$id))
  invisible(counter)
}
